aiosupervisor/client.py

```diff
@@ -0,0 +1,205 @@
+"""Internal client for making requests and managing session with Supervisor."""
+
+from dataclasses import dataclass, field
+from importlib import metadata
+from typing import Any
+
+from aiohttp import (
+    ClientError,
+    ClientResponse,
+    ClientResponseError,
+    ClientSession,
+    ClientTimeout,
+)
+from aiohttp.hdrs import METH_DELETE, METH_GET, METH_POST, METH_PUT
+from yarl import URL
+
+from .const import ResponseType
+from .exceptions import (
+    SupervisorAuthenticationError,
+    SupervisorBadRequestError,
+    SupervisorConnectionError,
+    SupervisorError,
+    SupervisorForbiddenError,
+    SupervisorNotFoundError,
+    SupervisorResponseError,
+    SupervisorServiceUnavailableError,
+    SupervisorTimeoutError,
+)
+from .models.base import Response, ResultType
+
+VERSION = metadata.version(__package__)
+
+
+def is_json(response: ClientResponse, raise_on_fail: bool = False) -> bool:
+    """Check if response is json according to Content-Type."""
+    content_type = response.headers.get("Content-Type", "")
+    if "application/json" not in content_type:
+        if raise_on_fail:
+            raise SupervisorError(
+                "Unexpected response received from supervisor when expecting"
+                f"JSON. Status: {response.status}, content type: {content_type}"
+            )
```

Comment:
It's used in two places. First one is in error handling here:
https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/aiosupervisor/client.py#L115

In this case I don't want it to raise if its not JSON because I have already set an exception type based on the status code. I want to raise that instead to give the client as much info as I can.

The second is if we got a successful response from supervisor but it wasn't JSON when we expected JSON:
https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/aiosupervisor/client.py#L122

In this case raising a generic `SupervisorError` is fine because this is already an unknown error situation. I choose to raise from the method as opposed to returning false and then raising because it felt cleaner to have the error message in one place (in case I used the function again) and I wouldn't have to duplicate the code to get the content type header.

Reply:
I didn't make any changes here, are we good with the way this works?

Reply:
> https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/aiosupervisor/client.py#L122
> 

IMHO, `is_json` should not have such side effects. `is_json` for me indicates a function which is very passive.


> I choose to raise from the method as opposed to returning false and then raising because it felt cleaner to have the error message in one place (in case I used the function again) and I wouldn't have to duplicate the code to get the content type header.

Yeah that is a very valid concern, and that is why I suggested two separate functions. This way we don't duplicate code, while still have function names which clearly state what the function is doing.

Anyhow, this is more stylistic, and internal anyways. I also don't care that much, we can leave as is if you prefer.